### PR TITLE
Kotlin/Native support

### LINKED ISSUE #80: Kotlin/Native TCP support
Since Ktor 1.4.0, CIO engine support TCP sockets (without TLS) on native. So it would be good to support it.
Firstly, we need to support Kotlin/Native, which will cause refactorings of some internals because of different memory model.
This will also allow to use Kotlin/Native with other transports later, when ktor will support them (f.e. websockets)

### PR DESCRIPTION
**Pull request by whyoleg**
* K/N support for linux, windows, macos, ios, watchos, tvos (fixes #80)
* TCP client and server support for all K/N targets supported by ktor (ktor doesn't support windows)
* Rewritten TCP connection implementation to not switch between threads while reading/writing (needed for K/N)
* Rewritten MPP `IntMap` implementation based on Netty `IntObjectHashMap` - on K/N it uses atomics to support modification, even if it's freezed. Implementation is synchronized, so it's possible to use it concurrently.

Should work with WS on K/N when ktor will support it (may be in 1.5.0).
Tests for almost all targets are running ~ 1 hour, except for tvos and watchos (up to 3 hours), where new K/N allocator isn't supported (yet?).

### CODE REVIEW

**Comment by yschimke** [comment]
No review - but this is epic!

**Comment by yschimke** [comment]
What does the integration story look like? e.g. is this a usable API for Swift or C++?

**Comment by whyoleg** [comment]
Not sure, that it will be easy to use it there because of suspend functions, but it possible, you can read it here: 
* https://kotlinlang.org/docs/reference/native/objc_interop.html
* https://kotlinlang.org/docs/reference/whatsnew14.html#support-for-kotlins-suspending-functions-in-swift-and-objective-c
IMO, it makes more sense to use it in common code in core shared logic.

**Comment by whyoleg** [comment]
PR is ready for review, but some changes in tests running on CI will land soon. As for now, some tests are running to long on CI

File: examples/multiplatform-chat/src/clientJsMain/kotlin/App.kt [hunk 1]
```diff
@@ -17,7 +17,7 @@
 import kotlinx.coroutines.flow.*
 
 suspend fun main() {
-    val api = connectToApi("Yuri")
+    val api = connectToApiUsingWS("Yuri")
```
Comment (yschimke) [inline examples/multiplatform-chat/src/clientJsMain/kotlin/App.kt:20]: Thanks... I guess.

**Review by yschimke — APPROVED** [review]
Review didn't seem too bad, mainly

1) build setup changes
2) TCP duplex connection using supported channels etc.
3) IntMap rewrite

File: examples/multiplatform-chat/src/clientJsMain/kotlin/App.kt (hunk 1 above)
Comment (whyoleg) [reply examples/multiplatform-chat/src/clientJsMain/kotlin/App.kt:20]: hah, yes, just first name which comes in mind after mine :)

**Pull request merged**